### ISSUE TITLE
Allow non-canonical type alignment?

### ISSUE THREAD

**Issue by abrown** [WebAssembly/component-model#258]
Over in wasi-nn, we have been wondering whether there might be a way to specify an alignment for a type larger than the one specified by the canonical ABI. This is similar to what @titzer brings up in https://github.com/WebAssembly/component-model/issues/194 but I thought it was different enough to merit its own issue.

wasi-nn currently specifies the tensor data as `list<u8>` ([WIT](https://github.com/WebAssembly/wasi-nn/blob/main/wit/wasi-nn.wit#L44)). The initial rationale (long ago, back in the WITX days) was a performance consideration: ML frameworks can deal with tensor data as byte blobs and if we can just "pass the bytes through" we can avoid some extra copies. This means that when we, as a wasi-nn guest user, receive tensor data back we must convert it from `list<u8>` to `list<f64>`. This conversion is an unsafe cast in some of the bindings &mdash; what if the `list<u8>` is not aligned? There has been an outstanding issue (https://github.com/WebAssembly/wasi-nn/issues/27) to figure this out. One solution would be to always copy the data; another to conditionally copy the data only if unaligned; etc.

The solution I'm asking about here is adding some way to inform the runtime of the alignment we would like when allocating space for the tensor data. Is this possible? Is it different enough from what @titzer is asking? I suspect we would only want to allow allocations for alignments larger than what the canonical ABI specifies, but maybe not.

**Comment by lukewagner** [comment]
In theory, yes, it makes sense and seems complementary to what we were talking about in #194, and maybe belongs as part of the same "feature".

Just to help me understand the motivation (and see if there's a short-term workaround): what's the reason for using a `list<u8>` instead of, say, a `list<u64>` (which would have essentially the same Canonical ABI as `list<u8>` except with the additional alignment check)?  That would mean having separately-named "overloads" for `list<u16>`-vs-`list<u32>`-vs-`list<u64>`; is that the problem?

**Comment by abrown** [comment]
Well, the `list<u64>` would "overallocate" when used with smaller types and certain unaligned sizes (e.g. passing 42 bytes of `u8` would result in 6 extra bytes allocated... with zeroes?). This is not really a huge problem but I was hoping for better. And having separate functions for each `list<*>` type is a bit wordy, but again not a huge problem. So, yes, that is the essence of the problem. There are a bunch of possible solutions here, some hackier than others, and what I'm hoping to figure out from this issue is where adding some extra syntax to WIT is a feasible option.

**Comment by lukewagner** [comment]
Yeah, I was asking about having a separate `list<T>` for each of `u8`, `u16`, `u32` and `u64`.  That would be the easy fix while we work on a more a full byte-layout-control mechanism.

**Comment by abrown** [comment]
Yeah, I don't think I'm in a terrible hurry so I can see two options: (a) use `list<T>` for all the types temporarily and then switch over or (b) just wait for the byte-layout-control mechanism. For (b), are you thinking about more than just an optional annotation to the [type definitions](https://github.com/WebAssembly/component-model/blob/main/design/mvp/Explainer.md#type-definitions), like `<defvaltype> ('align=' <align>)?` where `align ::= 'align=' 1 | 2 | 4 | 8 | 16`? It would need some way to say "don't allow alignment less than the minimum the canonical ABI specifies" or something like that, but I didn't see this addition as too complicated. Or maybe it is complicated by all the places it would need to be implemented in?

**Comment by lukewagner** [comment]
(sorry for the delay; back from the wasm CG meeting) That's a good question; thanks for asking.  Given that Preview 2 is almost done and there's not really time to make a significant addition like this, I think it's probably a good idea to go with approach (a), so that you can produce a nice working Preview 2 interface without waiting on anything else, and then we can include this issue as a use case for a possible Preview 3-or-after feature addition to the C-M/WIT that would improve the quality/ergonomics of the existing working interface.

**Comment by encounter** [comment]
I'd be interested to see a solution to this: my use case is parsing binary data from a `list<u8>`. If the binary data is, say, a 64-bit ELF, it's expected to be aligned to 8. When passing the data into the component, `cabi_realloc` is called with an alignment of 1 (and will end up with a minimum alignment of 4 on wasm32). I'd like to avoid reallocating inside of the component if the alignment ends up being lower than 8, and instead simply have the host align the allocation to 8 when it's passed in.

**Comment by lukewagner** [comment]
For the specific use case you mention @encounter, the lazy-lowering ABI addition proposed in #383 would address the issue by allowing guest code to control how allocations happen and, e.g., make an 8-byte-aligned allocation before lowering the `list<u8>` into it.  Technically this is also possible to do today with 0.2 `realloc` by supplying a custom `realloc` function the intentionally over-aligns, but it's a bit trickier to do in the toolchain.
